### PR TITLE
1119 smtp commands

### PR DESCRIPTION
**Pull request by pwitusik**

### CODE REVIEW

File: G1ANT.Addon.Net/Models/AttachmentModel.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+using MimeKit;
+
+namespace G1ANT.Addon.Net.Models
+{
+    public class AttachmentModel : IAttachmentModel
+    {
+        public string Name { get; }
+        public long Size { get; }
+        public string Type { get; }
+        public MimeEntity MimeEntity { get; }
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Models/AttachmentModel.cs]: I'm not sure if this should be public. Model should hide its internals.

File: G1ANT.Addon.Net/Models/AttachmentModel.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        public AttachmentModel(MimeEntity mimeEntity)
+        {
+            Name = GetAttachmentName(mimeEntity);
+            Size = GetAttachmentSize(mimeEntity);
+            Type = GetAttachmentMediaTypeName(mimeEntity);
+            MimeEntity = mimeEntity;
+        }
+
+        private string GetAttachmentName(MimeEntity mimeEntity)
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Models/AttachmentModel.cs]: MimeEntity is already a property of this object.
This class is a model. It should have state and should use it. It should not have methods that perform actions on - possibly - external data, because this is how services/business facades work.
TL;DR; remove arguments of all methods except constructor.

File: G1ANT.Addon.Net/Models/AttachmentModel.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        public MimeEntity MimeEntity { get; }
+
+        const string emptyAttachmentDefaultName = "unnamed-attachment";
+
+        public AttachmentModel(MimeEntity mimeEntity)
+        {
+            Name = GetAttachmentName(mimeEntity);
+            Size = GetAttachmentSize(mimeEntity);
+            Type = GetAttachmentMediaTypeName(mimeEntity);
+            MimeEntity = mimeEntity;
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Models/AttachmentModel.cs]: Setting object state should be its first activity, otherwise some methods called in constructor before object will be initialized might fail. This is also a pattern that might lead to bug introduction: you might expect that all fields/properties of the model are properly initialized during execution of your method, while in fact they are not.

File: G1ANT.Addon.Net/Models/AttachmentModel.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            {
+                fileName = emptyAttachmentDefaultName;
+            }
+            return fileName;
+        }
+
+        private long GetAttachmentSize(MimeEntity mimeEntity)
+        {
+            var size = mimeEntity.ContentDisposition?.Size.GetValueOrDefault(0);
+            return size ?? 0;
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Models/AttachmentModel.cs]: `GetValueOrDefault(0)` with another fallback to 0?

File: G1ANT.Addon.Net/Models/IAttachmentModel.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace G1ANT.Addon.Net.Models
+{
+
+    public interface IAttachmentModel
+    {
+        string Name { get; }
+        long Size { get; }
+        string Type { get; }
+        MimeEntity MimeEntity { get; }
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Models/IAttachmentModel.cs]: Nah...

File: G1ANT.Addon.Net/Models/IAttachmentModel.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+﻿using MimeKit;
+
+namespace G1ANT.Addon.Net.Models
+{
+
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Models/IAttachmentModel.cs]: Empty line (woo hoo!)

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 7]
```diff
@@ -1,27 +1,321 @@
...
-using G1ANT.Language;
+using System;
+using System.IO;
+using System.Linq;
+using System.Collections.Generic;
+using G1ANT.Addon.Net.Models;
 
 namespace G1ANT.Addon.Net
 {
-    public class SimplifiedMessageSummary : MessageSummary
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs:7]: So great that you removed this inheritance, because it looks it should be done in the opposite direction.
👍

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 8]
```diff
@@ -1,27 +1,321 @@
...
+        public List<object> Attachments
+        {
+            get => CreateAttachmentStructuresFromAttachments(FullMessage.Attachments);
+            set
+            {
+                var bodyBuilder = new BodyBuilder();
+                if (bodyBuilder.Attachments != null)
+                {
+                    bodyBuilder.HtmlBody = HtmlBody;
+                    bodyBuilder.TextBody = TextBody;
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs:57]: If there are no attachments unrelated properties (HtmlBody, TextBody, Body) won't be set. Looks weird for me.
Another weird thing is that you create `BodyBuilder` (line 53) without any parameters and one line below you expect that it will contain attachments. How is that possible? I'm not going to investigate how `BodyBuilder` works, I just want to show you that either lines 53-54 contain a bug or this method works in a weird way.

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 9]
```diff
@@ -1,27 +1,321 @@
...
+                    foreach (var attachment in value)
+                        bodyBuilder.Attachments.Add(attachment.ToString());
+                    FullMessage.Body = bodyBuilder.ToMessageBody();
+                }
+            }
+        }
+
+        private List<object> CreateAttachmentStructuresFromAttachments(IEnumerable<MimeEntity> attachments)
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs]: Why is this `List<object>` when you know that it is `List<AttachmentStructure>`?

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 10]
```diff
@@ -1,27 +1,321 @@
...
+        {
+            List<object> attachmentsList = new List<object>();
+
+            foreach (var attachment in attachments.Where(x => !string.IsNullOrEmpty(x.ContentDisposition?.FileName)))
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs]: Consider
```csharp
var attachmentsList = GetAttachmentsWithNamesSet(attachments)
    .Select(a => new AttachmentStructure(new AttachmentModel(a)))
    .ToList();
```

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 11]
```diff
@@ -1,27 +1,321 @@
...
+            {
+                AttachmentModel attachmentModel = new AttachmentModel(attachment);
+                AttachmentStructure temp = new AttachmentStructure(attachmentModel);
+                attachmentsList.Add(temp);
+            }
+            return attachmentsList;
+        }
+
+        public string MessageId
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs:77]: Nitpicking:
```csharp
public string MessageId => fullMessage?.MessageId ?? messageSummary?.Envelope.MessageId ?? string.Empty;
```
😈

Just joking, your method is way more readable.

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 12]
```diff
@@ -1,27 +1,321 @@
...
+        {
+            get
+            {
+                if (fullMessage != null)
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs:81]: There's something wrong with multiple null checks in every property. It looks like you try to make this model a container for two different objects. Consider adding an interface or abstract model and use it in two models that are containers just for one object (one for `MessageSummary`, one for `MimeMessage`). Or event don't bother with interface.
Otherwise you end up with lots of checks in lots of methods.

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 13]
```diff
@@ -1,27 +1,321 @@
...
+                    return fullMessage.MessageId;
+                else if (messageSummary != null)
+                    return messageSummary.Envelope.MessageId;
+                return string.Empty;
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs:85]: Throw exception in such case. But if you split this model into two separate ones then you will just do e.g. `public string MessageId => messageSummary.Envelope.MessageId;`. The code will simplify with each getter taking just one line without any logic inside.

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 14]
```diff
@@ -1,27 +1,321 @@
...
+        }
+
+        public MailboxAddress Sender
+        {
+            set
+            {
+                if (fullMessage != null)
+                    fullMessage.Sender = value;
+                else
+                    throw new NotSupportedException("Cannot set Sender for recieved message");
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs]: typo: received

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 15]
```diff
@@ -1,27 +1,321 @@
...
+                    return messageSummary.IsReply;
+                return false;
+            }
+        }
+
+        public string Priority
+        {
+            get
+            {
+                if (fullMessage != null && fullMessage.Headers != null)
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs]: I would say `if (fullMessage?.Headers != null)`, but if you rewrite this class into to separate ones a second part of existing condition will be enough.

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs [hunk 16]
```diff
@@ -1,27 +1,321 @@
...
+        private void FillBodyBuilderAttachments(BodyBuilder bodyBuilder)
+        {
+            if (bodyBuilder.Attachments != null)
+                foreach (var att in FullMessage.Attachments)
+                    bodyBuilder.Attachments.Add(att);
+        }
+
+        public string HtmlBody
+        {
+            get => FullMessage.HtmlBody;
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SimplifiedMessageSummary.cs:234]: Are we still in the same model, where once `fullMessage` is used and other time `messageSummary`?

Also, take a look: if you call `FullMessage` when `fullMessage` is null and `messageSummary` is not then you end up with both `fullMessage != null` and `messageSummary != null`. But a moment ago you initialized this class with `messageSummary` and `fullMessage` should be null. Now all getters return fields from different object.
What caused that was a simple call to `HtmlBody`. Come on, why does `HtmlBody` cause switch to another data source (summary=>full)? It's not called `GetHtmlBodyAndSwitchDataSourceToFullMessage`.

I read the code of whole class once again and I finally got your idea, but I'm still sure that this class should become two separate models + 3rd model that joins them together to achieve seamless/transparent conversion of data source.

File: G1ANT.Addon.Net/SmtpManager.cs [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+{
+    public sealed class SmtpManager
+    {
+        private NetworkCredential credentials;
+        private SmtpClient client;
+        private string host;
+        private int port;
+        private SecureSocketOptions socketOptions = SecureSocketOptions.Auto;
+
+        public static SmtpManager Instance { get; } = new SmtpManager();
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/SmtpManager.cs:20]: What if you want to use two instances at the same time? Consider Manager/Wrapper pattern as used in some other addons.

File: G1ANT.Addon.Net/Structures/AttachmentStructure.cs [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+                    return new IntegerStructure(Value.Size, null, Scripter);
+                case TypeIndex:
+                    return new TextStructure(Value.Type, null, Scripter);
+                case PathIndex:
+                    pathToTempFile = SaveAndGetPath(Value.MimeEntity, Value.Name);
+                    return new PathStructure(pathToTempFile, null, Scripter);
+                default:
+                    throw new ArgumentException($"Unknown index '{index}'", nameof(index));
+            }
+
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Structures/AttachmentStructure.cs]: Empty line  :P

File: G1ANT.Addon.Net/Structures/AttachmentStructure.cs [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        private string SaveAndGetPath(MimeEntity mimeEntity, string fileName)
+        {
+            var attachmentNameHash = GetAttachmentNameHash(fileName, mimeEntity.ContentId);
+            var filePath = GetAttachmentTempFileNamePath($"{attachmentNameHash}.{fileName}");
+            if (!File.Exists(filePath))
+            {
+                if (mimeEntity is MessagePart msgPart)
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Structures/AttachmentStructure.cs]: I'm not sure if structure should know such low-level internals.

File: G1ANT.Addon.Net/Structures/AttachmentStructure.cs [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+                    using (var stream = File.Create(filePath))
+                        msgPart.Message.WriteTo(stream);
+                }
+                else if (mimeEntity is MimePart mimePart)
+                {
+                    using (var stream = File.Create(filePath))
+                        mimePart.Content.DecodeTo(stream);
+                }
+                else
+                    return string.Empty;
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Structures/AttachmentStructure.cs]: Should throw an exception

File: G1ANT.Addon.Net/Structures/AttachmentStructure.cs [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+            }
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Structures/AttachmentStructure.cs]: Should throw an exception

File: G1ANT.Addon.Net/Structures/AttachmentStructure.cs [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            return tempFileService.GetTempPath(AttachmentFilePrefix, attachmentNameHash, "");
+        }
+
+        private string GetAttachmentNameHash(string fileName, string contentId)
+        {
+            return getMd5HashService.GetMd5Hash($"{contentId}-{fileName}");
+        }
+
+        private void DisposeAttachment(string attachmentPath)
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/Structures/AttachmentStructure.cs:70]: DisposeAttachment -> RemoveAttachmentFile

**Review by lfronczyk — CHANGES_REQUESTED** [review]

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs (hunk 8 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/SimplifiedMessageSummary.cs:57]: If Attachment property of BodyBuilder is null in some reason the attachments will not be included into the FullMessage.Body which can contain Text, HtmlText and Attachments of the mail message. It looks fine 4 me

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs (hunk 9 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/SimplifiedMessageSummary.cs]: Cause constructor of G1ANT's ListStructure accepts only List<object> :/

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs (hunk 10 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/SimplifiedMessageSummary.cs]: This code has not been modified when moving SimplifiedMessageSummary class from G1ANT.Language but ok 👍

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs (hunk 12 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/SimplifiedMessageSummary.cs:81]: I was thinking about that but...SimplifiedMessageSummary would have to wrap this interface and pass calls to it, depending on level of data needed it should then manage switching between different implementations of the interface.
It's too much hassle

File: G1ANT.Addon.Net/SimplifiedMessageSummary.cs (hunk 16 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/SimplifiedMessageSummary.cs:234]: If we want to make it right we should also rewrite smtp commands which retrieve emails. In general it's to much time to spent to refactor and rewrite everything

File: G1ANT.Addon.Net/SmtpManager.cs (hunk 17 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/SmtpManager.cs:20]: It has been developed in this way months ago. I'm not going to refactor whole addon :P

File: G1ANT.Addon.Net/Structures/AttachmentStructure.cs (hunk 22 above)
Comment (pwitusik) [reply G1ANT.Addon.Net/Structures/AttachmentStructure.cs:70]: This code has been moved from G1ANT.Language, didn't you notice that? It is the name from original class

**Review by lfronczyk — APPROVED** [review]
